src/main/java/org/opensearch/flowframework/workflow/AbstractRetryableWorkflowStep.java

```diff
@@ -65,82 +70,92 @@ protected AbstractRetryableWorkflowStep(
-                    } else {
-                        id = response.getTaskId();
-                    }
-                    flowFrameworkIndicesHandler.updateResourceInStateIndex(
-                        workflowId,
-                        nodeId,
-                        getName(),
-                        id,
-                        ActionListener.wrap(updateResponse -> {
-                            logger.info("successfully updated resources created in state index: {}", updateResponse.getIndex());
-                            future.complete(
-                                new WorkflowData(
-                                    Map.ofEntries(
-                                        Map.entry(resourceName, response.getModelId()),
-                                        Map.entry(REGISTER_MODEL_STATUS, response.getState().name())
-                                    ),
-                                    workflowId,
-                                    nodeId
-                                )
-                            );
-                        }, exception -> {
-                            logger.error("Failed to update new created resource", exception);
-                            future.completeExceptionally(
-                                new FlowFrameworkException(exception.getMessage(), ExceptionsHelper.status(exception))
-                            );
-                        })
-                    );
-                } catch (Exception e) {
-                    logger.error("Failed to parse and update new created resource", e);
-                    future.completeExceptionally(new FlowFrameworkException(e.getMessage(), ExceptionsHelper.status(e)));
-                }
-            }
-        }, exception -> {
-            if (retries < maxRetry) {
-                // Sleep thread prior to retrying request
+                }));
+                // Wait long enough for future to possibly complete
                 try {
                     Thread.sleep(5000);
-                } catch (Exception e) {
+                } catch (InterruptedException e) {
                     FutureUtils.cancel(future);
+                    Thread.currentThread().interrupt();
                 }
-                retryableGetMlTask(workflowId, nodeId, future, taskId, retries + 1, workflowStep);
-            } else {
-                logger.error("Failed to retrieve" + workflowStep + ",maximum retries exceeded");
-                future.completeExceptionally(new FlowFrameworkException(exception.getMessage(), ExceptionsHelper.status(exception)));
             }
-        }));
+            if (!future.isDone()) {
+                String errorMessage = workflowStep + " did not complete after " + maxRetry + " retries";
+                logger.error(errorMessage);
+                future.completeExceptionally(new FlowFrameworkException(errorMessage, RestStatus.REQUEST_TIMEOUT));
+            }
+        }, threadPool.executor(PROVISION_THREAD_POOL));
```

Comment:
just for learning purpose, why do we need this exactly now, when we didn't have this before (the `threadPool.executor`

Reply:
Previously we were using an ActionListener (either `onResponse` or `onFailure`) to respond to the result, and relying on an exception to terminate that listener thread.   The new implementation uses `runAsync()` which needs an executor.  We could use the OpenSearch `generic()` one but since we've defined our own provision thread pool and this retry is part of provisioning it seemed the best one to use for this purpose.

Reply:
that sounds good to me, I am not sure on the 100% best standard of using threads in this way from within the plugin, but relaying on our provision thread pool makes sense to me